File: 2019/04/perl/04.pl
#!/usr/bin/env perl
use List::Util qw/sum/;

#global regex
$non_decrease = '^(?=\d{6}$)1*2*3*4*5*6*7*8*9*$';
$repeat       = '(\d)\1';

sub p1_valid
   {
   1 if ( $_ =~ /$non_decrease/ and $_ =~ /$repeat/ );
   }

sub p2_valid
   {
   #first check non decreasing
   my $cond1  = $_ =~ /$non_decrease/;

   #remove repeats of more than length 2
   $_ =~ s/(\d)\1{2,}//g;

   #now check for repeats
   my $cond2 = $_ =~ /$repeat/;

   1 if ( $cond1 and $cond2 );
   }

sub main
   {
   my @input = (264360..746325);

   my @p1 = map { p1_valid($_) } @input;
   my @p2 = map { p2_valid($_) } @input;

   print "Part 1 answer: " . (sum @p1) . "\n";
   print "Part 2 answer: " . (sum @p2) . "\n";

   }

main();

File: 2019/04/python/04.py
#!/usr/bin/env python
from itertools import groupby

def valid(passwd, mode):
   #check adjacent characters
   non_decrease = all([a <= b for a, b in zip(passwd, passwd[1:])])

   #get repeated characters
   lengths = [len(''.join(g)) for _, g in groupby(passwd)]

   if mode == 'p1':
      repeat = max(lengths) >= 2
   elif mode == 'p2':
      repeat = 2 in lengths
   else:
      raise ValueError

   return (repeat and non_decrease)

if __name__ == '__main__':
   input_list = map(str, range(264360, 746325+1))
   input_list = list(input_list)

   p1 = [passwd for passwd in input_list if valid(passwd, 'p1')]
   p2 = [passwd for passwd in input_list if valid(passwd, 'p2')]

   print(f"Part 1 answer: {len(p1)}")
   print(f"Part 2 answer: {len(p2)}")
